File: backend/routers/analytics.py
import re
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from backend.db.engine import get_db
from backend.repositories.analytics_repo import (
    get_account_monthly_summary,
    get_grade_benchmarks,
    get_overview,
    get_follower_growth,
    get_top_profiles,
    get_hashtag_frequency,
    get_hashtag_performance,
    get_engagement_by_profile,
    get_post_engagement_history,
    get_post_volume,
    get_posting_time_heatmap,
    get_scrape_run_summary,
    search_similar_posts,
    get_wisdom_warriors_monthly_views_filtered,
)

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


@router.get("/overview")
async def overview(db: AsyncSession = Depends(get_db)) -> dict:
    return await get_overview(db)


@router.get("/follower-growth")
async def follower_growth(username: str | None = None, db: AsyncSession = Depends(get_db)) -> list:
    return await get_follower_growth(db, username)


@router.get("/top-profiles")
async def top_profiles(
    metric: str = "followers_count", limit: int = 10, db: AsyncSession = Depends(get_db)
) -> list:
    return await get_top_profiles(db, metric, limit)


@router.get("/hashtag-frequency")
async def hashtag_frequency(limit: int = 20, db: AsyncSession = Depends(get_db)) -> list:
    return await get_hashtag_frequency(db, limit)


@router.get("/engagement-by-profile")
async def engagement(db: AsyncSession = Depends(get_db)) -> list:
    return await get_engagement_by_profile(db)


@router.get("/post-trends")
async def post_trends(db: AsyncSession = Depends(get_db)) -> list:
    return await get_post_volume(db)


@router.get("/account-summary")
async def account_summary(
    period_label: str | None = None,
    limit: int = 12,
    db: AsyncSession = Depends(get_db),
) -> list:
    return await get_account_monthly_summary(db, period_label, limit)


@router.get("/grade-benchmarks")
async def grade_benchmarks(
    period_label: str | None = None,
    db: AsyncSession = Depends(get_db),
) -> list:
    return await get_grade_benchmarks(db, period_label)


@router.get("/hashtag-performance")
async def hashtag_performance(
    period_label: str | None = None,
    username: str | None = None,
    limit: int = 15,
    db: AsyncSession = Depends(get_db),
) -> list:
    return await get_hashtag_performance(db, period_label, username, limit)


@router.get("/posting-time-heatmap")
async def posting_time_heatmap(
    username: str | None = None,
    db: AsyncSession = Depends(get_db),
) -> list:
    return await get_posting_time_heatmap(db, username)


@router.get("/scrape-run-summary")
async def scrape_run_summary(limit: int = 10, db: AsyncSession = Depends(get_db)) -> list:
    return await get_scrape_run_summary(db, limit)


@router.get("/post-engagement-history")
async def post_engagement_history(short_code: str, db: AsyncSession = Depends(get_db)) -> list:
    return await get_post_engagement_history(db, short_code)


@router.get("/semantic-post-search")
async def semantic_post_search(
    query: str,
    username: str | None = None,
    limit: int = 10,
    db: AsyncSession = Depends(get_db),
) -> list:
    query = query.strip()
    if not query:
        raise HTTPException(status_code=400, detail="query is required")
    return await search_similar_posts(db, query=query, username=username, limit=limit)


@router.get("/wisdom-warriors/monthly-views")
async def wisdom_warriors_monthly_views(
    month: str,
    apply_filters: bool = True,
    category: str | None = None,
    hashtags: list[str] | None = Query(default=None),
    mentions: list[str] | None = Query(default=None),
    keywords: list[str] | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
) -> list:
    if not re.fullmatch(r"\d{4}-\d{2}", month):
        raise HTTPException(status_code=400, detail="month must be in YYYY-MM format")
    return await get_wisdom_warriors_monthly_views_filtered(
        db=db,
        month=month,
        apply_filters=apply_filters,
        hashtags=hashtags,
        mentions=mentions,
        caption_keywords=keywords,
        category=category,
    )

File: frontend/src/api/analytics.ts
import { API_URL } from "../config"
import type {
  OverviewStats,
  TimeSeriesPoint,
  TopProfile,
  HashtagFrequency,
  EngagementByProfile,
  PostVolume,
  AccountSummary,
  GradeBenchmark,
  HashtagPerformance,
  PostingTimeHeatmapPoint,
  ScrapeRunSummary,
  SemanticPostResult,
  PostEngagementHistoryPoint,
} from "../types/analytics"


const buildQueryString = (params: Record<string, string | number | undefined>) => {
  const searchParams = new URLSearchParams()
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== "") {
      searchParams.set(key, String(value))
    }
  }
  const query = searchParams.toString()
  return query ? `?${query}` : ""
}

export const fetchOverview = (periodLabel?: string): Promise<OverviewStats> =>
  fetch(`${API_URL}/api/analytics/overview${buildQueryString({ period_label: periodLabel })}`).then(r => r.json())

export const fetchFollowerGrowth = (username?: string, upToPeriodLabel?: string): Promise<TimeSeriesPoint[]> =>
  fetch(`${API_URL}/api/analytics/follower-growth${buildQueryString({ username, up_to_period_label: upToPeriodLabel })}`).then(r => r.json())

export const fetchTopProfiles = (metric = "followers_count", limit = 10): Promise<TopProfile[]> =>
  fetch(`${API_URL}/api/analytics/top-profiles?metric=${metric}&limit=${limit}`).then(r => r.json())

export const fetchHashtagFrequency = (limit = 20): Promise<HashtagFrequency[]> =>
  fetch(`${API_URL}/api/analytics/hashtag-frequency?limit=${limit}`).then(r => r.json())

export const fetchEngagement = (): Promise<EngagementByProfile[]> =>
  fetch(`${API_URL}/api/analytics/engagement-by-profile`).then(r => r.json())

export const fetchPostVolume = (upToPeriodLabel?: string): Promise<PostVolume[]> =>
  fetch(`${API_URL}/api/analytics/post-trends${buildQueryString({ up_to_period_label: upToPeriodLabel })}`).then(r => r.json())

export const fetchAccountSummary = (periodLabel?: string, limit = 12): Promise<AccountSummary[]> =>
  fetch(`${API_URL}/api/analytics/account-summary${buildQueryString({ period_label: periodLabel, limit })}`).then(r => r.json())

export const fetchGradeBenchmarks = (periodLabel?: string): Promise<GradeBenchmark[]> =>
  fetch(`${API_URL}/api/analytics/grade-benchmarks${buildQueryString({ period_label: periodLabel })}`).then(r => r.json())

export const fetchHashtagPerformance = (periodLabel?: string, username?: string, limit = 12): Promise<HashtagPerformance[]> =>
  fetch(`${API_URL}/api/analytics/hashtag-performance${buildQueryString({ period_label: periodLabel, username, limit })}`).then(r => r.json())

export const fetchPostingTimeHeatmap = (username?: string, periodLabel?: string): Promise<PostingTimeHeatmapPoint[]> =>
  fetch(`${API_URL}/api/analytics/posting-time-heatmap${buildQueryString({ username, period_label: periodLabel })}`).then(r => r.json())

export const fetchScrapeRunSummary = (limit = 8, maxRunId?: number): Promise<ScrapeRunSummary[]> =>
  fetch(`${API_URL}/api/analytics/scrape-run-summary${buildQueryString({ limit, max_run_id: maxRunId })}`).then(r => r.json())

export const fetchSemanticPostSearch = (query: string, username?: string, limit = 8): Promise<SemanticPostResult[]> =>
  fetch(`${API_URL}/api/analytics/semantic-post-search${buildQueryString({ query, username, limit })}`).then(async r => {
    if (!r.ok) {
      const err = await r.json().catch(() => ({ detail: "Failed to search posts" }))
      throw new Error(err.detail ?? "Failed to search posts")
    }
    return r.json()
  })

export const fetchPostEngagementHistory = (shortCode: string): Promise<PostEngagementHistoryPoint[]> =>
  fetch(`${API_URL}/api/analytics/post-engagement-history${buildQueryString({ short_code: shortCode })}`).then(r => r.json())
